### PR TITLE
[FORMS-9062] improve accessibility

### PR DESCRIPTION
**Pull request by amit-mnnit**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes and the overall coverage did not decrease.
- [ ] All unit tests pass on CircleCi.
- [ ] I ran all tests locally and they pass.

### CODE REVIEW

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/af-commons/v1/fieldTemplates/label.html [hunk 1]
```diff
@@ -1,4 +1,7 @@
-<template data-sly-template.label="${@ componentId, labelValue, labelVisible, bemBlock}">
+<template data-sly-template.label="${@ componentId, labelValue, labelVisible, bemBlock, withId}">
```
Comment (deepprakash345) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/af-commons/v1/fieldTemplates/label.html:1]: can we use a better terminology, withId is not intuitive.

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/button.html [hunk 2]
```diff
@@ -34,9 +34,9 @@
       disabled="${!button.enabled}"
       title="${button.tooltipVisible ? '' :   button.tooltipText}"
       data-cmp-data-layer="${button.data.json}"
-      aria-label="${button.label.value}">
+      aria-label="${button.label.visible != false ? '' : button.label.value}">
```
Comment (deepprakash345) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/button/v1/button/button.html:37]: Please add test case for this change, we already saw test failures because of this.

Add test case for both scenarios.
1. when cq dialog is not openend.
2. when cq dialog is openend and saved

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/checkboxgroup/v1/checkboxgroup/clientlibs/site/js/checkboxgroupview.js [hunk 3]
```diff
@@ -130,7 +130,6 @@
         }
 
         updateReadOnly(readonly) {
-            this.toggle(readonly, "aria-readonly", true);
             let widgets = this.widget;
```
Comment (deepprakash345) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/checkboxgroup/v1/checkboxgroup/clientlibs/site/js/checkboxgroupview.js:133]: Is this removal intentional??

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/datepicker.html [hunk 4]
```diff
@@ -38,6 +39,8 @@
             required="${datePicker.required}"
             data-cmp-data-layer="${datePicker.data.json}"
             title="${datePicker.tooltipVisible ? '' : datePicker.tooltipText}"
+            aria-label="${datePicker.label.visible ? '' : datePicker.label.value}"
+            id="${datePicker.label.visible ? widgetId : ''}"
```
Comment (deepprakash345) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/datepicker.html:43]: why is presence of id field of datepicker is depending on presence of label

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/emailinput/v1/emailinput/emailinput.html [hunk 5]
```diff
@@ -21,18 +21,20 @@
      data-sly-use.shortDescription="${renderer.shortDescriptionPath}"
      data-sly-use.longDescription="${renderer.longDescriptionPath}"
      data-sly-use.errorMessage="${renderer.errorMessagePath}"
-     data-sly-use.questionMark="${renderer.questionMarkPath}"></sly>
+     data-sly-use.questionMark="${renderer.questionMarkPath}"
+     data-sly-test.widgetId="${'{0}-{1}' @ format=[email.id, 'widget']}"></sly>
```
Comment (deepprakash345) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/emailinput/v1/emailinput/emailinput.html:25]: if we are putting in if condition to determine rendering, we must add testcases for the scenarios to ensure no future regressions

**Review by deepprakash345 — DISMISSED** [review]
-1 for now. 
Please add test cases as per the comments inline

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/emailinput/v1/emailinput/emailinput.html (hunk 5 above)
Comment (amit-mnnit) [reply ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/emailinput/v1/emailinput/emailinput.html:25]: @deepprakash345 
There are known test case failures. I have changed the value of the “for” attribute for all components. Once the PR is approved, I will update the test cases and add new test cases.

**Comment by deepprakash345** [comment]
@amit-mnnit , can you take dev merge and run the tests again

**Comment by amit-mnnit** [comment]
> @amit-mnnit , can you take dev merge and run the tests again

donr @deepprakash345

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/panelcontainer/v1/panelcontainer/simple.html [hunk 6]
```diff
@@ -24,8 +24,10 @@
          class="cmp-container${wcmmode.edit ? ' {0}': '' @ format=[allowed.cssClass]}"
          title="${panel.tooltipVisible ? '' : panel.tooltipText}"
          role="${container.roleAttribute}"
-         style="${container.backgroundStyle @ context='styleString'}">
-        <div data-sly-call="${label.label @componentId=panel.id, labelValue=panel.label.value, labelVisible=panel.label.visible, bemBlock='cmp-container'}" data-sly-unwrap></div>
+         style="${container.backgroundStyle @ context='styleString'}"
+         data-sly-test.widgetId="${'{0}-{1}' @ format=[panel.id, 'widget']}"
+         aria-labelledby="${panel.label.visible != false ? widgetId : ''}">
+        <div data-sly-call="${label.label @componentId=widgetId, labelValue=panel.label.value, labelVisible=panel.label.visible, bemBlock='cmp-container', withId=true}" data-sly-unwrap></div>
```
Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/panelcontainer/v1/panelcontainer/simple.html:30]: I don't see the value of withId set to false anywhere ? Is it true always

Comment (amit-mnnit) [reply ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/panelcontainer/v1/panelcontainer/simple.html:30]: the default value of withId is false

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/af-commons/v1/fieldTemplates/label.html [hunk 7]
```diff
@@ -1,4 +1,7 @@
...
     <label for="${componentId}" class="${bemBlock}__label"
-           data-sly-test="${labelValue && (labelVisible==null || labelVisible==true)}">${labelValue}</label>
+           data-sly-test="${!withId && labelValue && (labelVisible==null || labelVisible==true)}">${labelValue}</label>
+
+    <label id="${componentId}" class="${bemBlock}__label"
+           data-sly-test="${withId && labelValue && (labelVisible==null || labelVisible==true)}">${labelValue}</label>
```
Comment (sakshi-arora1) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/af-commons/v1/fieldTemplates/label.html:6]: can we test add attributes depending on sly-test, instead of repeating the label element.

**Comment by nit23uec** [comment]
as discussed, we should avoid id's in the markup until absolutely necessary. So, we can avoid giving id's to widgets.
the current `<label for="componentId">` should suffice for most of the accessibility requirements as id there is the id of the component that we already have.
+1 to introducing `aria-label` in case label is hidden.
minor: for question mark, title could be "Help Text" as it resonates with the user friendly name of this feature and is consistent with the naming conventions followed in edit dialogs as well.
Ideally, title attribute is not a good way for hover text (short description) as it's not accessibility friendly on touch enabled devices. we should use `aria-describedby` along with `role="tooltip"` for such cases.
for removing `id=emptyValue` pls discuss it with @ravise5 if there is any functionality impact  as he gave support for empty options in dropdown, IIRC.

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/telephoneinput/v1/telephoneinput/telephoneinput.html [hunk 8]
```diff
@@ -32,7 +32,7 @@
     <div data-sly-call="${label.label @componentId=telText.id, labelValue=telText.label.value, labelVisible=telText.label.visible, bemBlock='cmp-adaptiveform-telephoneinput'}" data-sly-unwrap></div>
     <input title="${telText.tooltipVisible ? '' : telText.tooltipText}"
            class="cmp-adaptiveform-telephoneinput__widget"
-           aria-label="${telText.label.visible && telText.label.value}"
+           aria-label="${telText.label.value}"
```
Comment (nit23uec) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/telephoneinput/v1/telephoneinput/telephoneinput.html:35]: why aria-label is required when it can be taken care by the `label for `itself.

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/textinput/v1/textinput/textinput.html [hunk 9]
```diff
@@ -32,7 +32,7 @@
     <div data-sly-call="${label.label @componentId=text.id, labelValue=text.label.value, labelVisible=text.label.visible, bemBlock='cmp-adaptiveform-textinput'}" data-sly-unwrap></div>
     <textarea data-sly-test.useTextarea="${text.multiLine}"
               title="${text.tooltipVisible ? '' : text.tooltipText}"
-              aria-label="${text.label.visible && text.label.value}"
+              aria-label="${text.label.value}"
```
Comment (nit23uec) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/textinput/v1/textinput/textinput.html:35]: any reasons for this change?

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.html [hunk 10]
```diff
@@ -30,10 +30,10 @@
      id="${numberinput.id}"
      data-cmp-adaptiveformcontainer-path="${formstructparser.formContainerPath}">
     <div data-sly-call="${label.label @componentId=numberinput.id, labelValue=numberinput.label.value, labelVisible=numberinput.label.visible, bemBlock='cmp-adaptiveform-numberinput'}" data-sly-unwrap></div>
-    <div data-sly-call="${questionMark.questionMark @componentId=numberinput.id, longDescription=numberinput.description, bemBlock='cmp-adaptiveform-numberinput'}"></div>
+    <div data-sly-call="${questionMark.questionMark @componentId=numberinput.id, longDescription=numberinput.description, bemBlock='cmp-adaptiveform-numberinput'}" data-sly-unwrap></div>
     <input
             class="cmp-adaptiveform-numberinput__widget"
-            aria-label="${numberinput.label.visible && numberinput.label.value}"
+            aria-label="${numberinput.label.value}"
```
Comment (nit23uec) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.html:36]: any reasons for this change?

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/telephoneinput/v1/telephoneinput/telephoneinput.html (hunk 8 above)
Comment (amit-mnnit) [reply ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/telephoneinput/v1/telephoneinput/telephoneinput.html:35]: @nit23uec 
Screenshot 2023-06-29 at 4 35 42 PM

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/fileinput/v1/fileinput/fileinput.html [hunk 11]
```diff
@@ -42,6 +42,7 @@
            accept="${file.accept}"
            id="${file.id}_widget"
            type="file"
+           aria-label="${file.label.value}"
```
Comment (nit23uec) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/fileinput/v1/fileinput/fileinput.html:45]: rich text wont be supported here in accessibility

Comment (rismehta) [reply ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/fileinput/v1/fileinput/fileinput.html:45]: Screen readers are designed to provide auditory feedback to users with visual impairments, and they can recognize and interpret some HTML tags within certain contexts.

**Review by rismehta — APPROVED** [review]
@amit-mnnit Please incorporate other comments iteratively, merging this

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/fileinput/v1/fileinput/fileinput.html (hunk 11 above)
Comment (amit-mnnit) [reply ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/fileinput/v1/fileinput/fileinput.html]: > rich text wont be supported here in accessibility

Removed

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/fileinput/v1/fileinput/clientlibs/site/js/fileinputview.js [hunk 12]
```diff
@@ -99,9 +99,18 @@
 
         }
 
+        // add comment
```
[the remark this answers is not in the document]
Comment (amit-mnnit) [reply ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/fileinput/v1/fileinput/clientlibs/site/js/fileinputview.js]: done

**Review by nit23uec — APPROVED** [review]

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/fileinput/v1/fileinput/fileinput.html (hunk 11 above)
Comment (nit23uec) [reply ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/fileinput/v1/fileinput/fileinput.html]: > they can recognize and interpret some HTML tags within certain contexts
only when it is not string. this does not holds true in this context as this is a string.

**Review by nit23uec — APPROVED** [review]
